File: packages/ts/src/manifest.ts
// Loads the canonical rules-manifest.json from the repo root.
// This is the SINGLE source of truth shared with the Rust surface (include_str!).
// Path note: this file lives at packages/ts/src/, so ../../../ = the package repo root
// (jsonstat-validator/) whether run from src via tsx or from dist via a tsc build.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import type { ManifestData, ManifestRule } from "./types.js";

const here = path.dirname(fileURLToPath(import.meta.url));
const manifestPath = path.resolve(here, "../../../rules-manifest.json");

export const MANIFEST: ManifestData = JSON.parse(
  readFileSync(manifestPath, "utf8"),
);

const byId = new Map<string, ManifestRule>(MANIFEST.rules.map((r) => [r.id, r]));

export function lookupRule(id: string): ManifestRule {
  const r = byId.get(id);
  if (!r) throw new Error(`jsonstat-validator: unknown rule id "${id}"`);
  return r;
}

File: crates/validator/src/manifest.rs
pub use crate::types::{ManifestData, ManifestRule};

/// The single source of truth, shared verbatim with the TypeScript surface.
/// Embedded from the copy synced into src/_vendored/manifest/ by build.rs. build.rs copies it from
/// the monorepo-root ../../rules-manifest.json (the shared single source of truth) so the crate stays
/// self-contained for `cargo publish` (which only packs files inside the package directory).
pub const MANIFEST_JSON: &str = include_str!("_vendored/manifest/rules-manifest.json");

pub fn manifest() -> ManifestData {
    serde_json::from_str(MANIFEST_JSON)
        .expect("rules-manifest.json must be valid JSON matching ManifestData")
}

pub fn lookup<'a>(mf: &'a ManifestData, id: &str) -> &'a ManifestRule {
    mf.rules
        .iter()
        .find(|r| r.id == id)
        .unwrap_or_else(|| panic!("jsonstat-validator: unknown rule id \"{id}\""))
}
